### DISCUSSION TITLE
Add new key overlay type

### DISCUSSION THREAD

**Discussion by Terochi** [ppy/osu#22602] [Ideas]
I think it would be nice to have a visual key overlay similar to the one most streamers use (https://github.com/Blondazz/KeyOverlay)

I am not sure about the design, but I've started with some basic implementation.
https://github.com/ppy/osu/compare/master...Terochi:KeyOverlay
![image](https://user-images.githubusercontent.com/100224757/218196371-0925515a-d030-422d-9813-195899c07ef6.png)

As some of its extra features I propose having each block colored based on combo color or red color to signify miss.
A nice touch would also be showing "autopilot key presses" as an optional overlap.

**Comment by mk56-spn** [comment]
Heads up. I was going to do this but never got around to it because ideally the keycounter would be refactored into a skinnable drawable first before adding more stuff. And doing so proved to be a tremendous pain in the rear given how it is currently structured . Maybe you're better than me at programming and this ends up being a non issue for you but thought you should know that

**Reply by Terochi** [reply]
I will give it a try, however I'm wondering whether I should make a container with all binded keys inside it with an array like `[SettingSource]` to configure which keys should be shown or just make the user add more components for each individual key.

**Reply by mk56-spn** [reply]
I'm not sure if we want that degree of customisability but if it is added I'm very much inclined to think that `[SettingSource]`is the way forward. Seems much cleaner than individual keys. Although I'd probably wait for bdach or peppy to pop in here and give clearer instructions wrt direction

**Reply by Terochi** [reply]
I managed to somewhat make the `KeyCounter` an `ISkinnableDrawable`, but couldn't fix the errors in tests that appeared because of that change. I'm gonna leave this idea be if someone wants to try as well.

**Reply by mk56-spn** [reply]
> I managed to somewhat make the `KeyCounter` an `ISkinnableDrawable`, but couldn't fix the errors in tests that appeared because of that change. I'm gonna leave this idea be if someone wants to try as well.

Mind sharing the code? might give it a shot if youve gotten past the things that blocked me, will credit you ofc if i PR with your efforts

**Reply by Terochi** [reply]
Sure, here you go
https://github.com/ppy/osu/compare/master...Terochi:osu:key-counter-skinnable-drawable
It isn't the cleanest, as I'm for example still using `ICanAttachHUDPieces` even though it is not a HUD piece anymore
